newton/_src/sensors/frame_transform_sensor.py

```diff
@@ -0,0 +1,232 @@
+    if body_idx >= 0:
+        # Shape attached to a body
+        X_wb = body_q[body_idx]
+        X_bs = shape_transform[shape_idx]
+        world_transforms[shape_idx] = wp.transform_multiply(X_wb, X_bs)
+    else:
+        # Static shape in world frame
+        world_transforms[shape_idx] = shape_transform[shape_idx]
+
+
+@wp.kernel
+def compute_relative_transforms_kernel(
+    all_shape_transforms: wp.array(dtype=wp.transform),
+    shape_indices: wp.array(dtype=int),
+    reference_indices: wp.array(dtype=int),
+    # output
+    relative_transforms: wp.array(dtype=wp.transform),
+):
+    """Compute relative transforms expressing object poses in reference frame coordinates.
+
+    Args:
+        all_shape_transforms: Array of world transforms for all shapes (indexed by shape index)
+        shape_indices: Indices of target shapes
+        reference_indices: Indices of reference sites
+        relative_transforms: Output array of relative transforms
+
+    Computes X_ro = X_wr^{-1} * X_wo for each pair, where:
+    - X_wo is the world transform of the object shape (object to world)
+    - X_wr is the world transform of the reference site (reference to world)
+    - X_ro is the transform from object to reference (expresses object pose in reference frame)
+    """
+    tid = wp.tid()
+    shape_idx = shape_indices[tid]
+    ref_idx = reference_indices[tid]
+
+    X_wo = all_shape_transforms[shape_idx]
+    X_wr = all_shape_transforms[ref_idx]
+
+    # Compute relative transform: express object pose in reference frame coordinates
+    X_ro = wp.transform_multiply(wp.transform_inverse(X_wr), X_wo)
+    relative_transforms[tid] = X_ro
+
+
+class FrameTransformSensor:
+    """Sensor that measures transforms of shapes/sites relative to reference sites.
+
+    This sensor computes the transform from a reference frame (site) to target shapes
+    (which can be regular shapes or sites).
+
+    Attributes:
+        shape_indices: Indices of shapes to measure
+        reference_site_indices: Indices of reference sites (one per shape or one shared)
+        transforms: Output array of relative transforms (updated after each call to update())
+
+    Example:
+        Measure shapes relative to a site::
+
+            # Get shape indices somehow (e.g., via selection or direct indexing)
+            shape_indices = [0, 1, 2]  # indices of shapes to measure
+            reference_site_idx = 5  # index of reference site
+
+            sensor = FrameTransformSensor(
+                model,
+                shape_indices=shape_indices,
+                reference_site_indices=[reference_site_idx],
+            )
+
+            # Update after eval_fk
+            sensor.update(model, state)
+
+            # Access transforms
+            transforms = sensor.transforms.numpy()  # shape: (N, 7) [pos, quat]
+    """
+
+    def __init__(
+        self,
+        model: Model,
+        shape_indices: list[int],
+        reference_site_indices: list[int],
```

Comment:
Consider naming those arguments `shapes` and `reference_sites`. That those are indices should be clear (we have those assumptions in all the ModelBuilder methods as well).